File: skills/references/channels.js
/**
 * Pattern: Channels
 * Comparable to: Unix pipes, gRPC streaming, WebSocket data streams
 *
 * Demonstrates binary streaming between workers: creating channels,
 * passing refs across functions, writing/reading binary data, and
 * using text messages for signaling.
 *
 * How-to references:
 *   - Channels: https://iii.dev/docs/how-to/use-channels
 */

import { registerWorker, Logger } from 'iii-sdk'

const iii = registerWorker(process.env.III_ENGINE_URL || 'ws://localhost:49134', {
  workerName: 'channels-example',
})

// ---------------------------------------------------------------------------
// 1. Producer — creates a channel and streams data through it
// ---------------------------------------------------------------------------
iii.registerFunction('pipeline::produce', async (data) => {
  const logger = new Logger()

  // Create a channel pair
  const channel = await iii.createChannel()

  // Pass the reader ref to the consumer via trigger
  iii.trigger({
    function_id: 'pipeline::consume',
    payload: {
      readerRef: channel.readerRef,
      recordCount: data.records.length,
    },
  })

  // Send metadata as a text message
  channel.writer.sendMessage(
    JSON.stringify({ type: 'metadata', format: 'ndjson', encoding: 'utf-8' }),
  )

  // Stream records as binary data (newline-delimited JSON)
  for (const record of data.records) {
    const line = JSON.stringify(record) + '\n'
    channel.writer.stream.write(Buffer.from(line))
  }

  // Signal end of stream
  channel.writer.close()
  logger.info('Producer finished streaming', { records: data.records.length })

  return { status: 'streaming', readerRef: channel.readerRef }
})

// ---------------------------------------------------------------------------
// 2. Consumer — receives a channel ref and reads the stream
// ---------------------------------------------------------------------------
iii.registerFunction('pipeline::consume', async (data) => {
  const logger = new Logger()

  // Reconstruct reader from the ref passed in the payload
  const reader = data.readerRef

  // Listen for text messages (metadata, signaling)
  reader.onMessage((msg) => {
    const parsed = JSON.parse(msg)
    logger.info('Received metadata', parsed)
  })

  // Read entire binary stream
  const buffer = await reader.readAll()
  const text = buffer.toString('utf-8').trim()
  
  let records
  if (!text) {
    records = [{ processed: 0 }]
  } else {
    const lines = text.split('\n').filter((line) => line.trim() !== '')
    records = lines.map((line) => JSON.parse(line))
  }

  logger.info('Consumer processed records', { count: records.length })
  return { processed: records.length }
})

// ---------------------------------------------------------------------------
// 3. HTTP trigger to kick off the pipeline
// ---------------------------------------------------------------------------
iii.registerTrigger({
  type: 'http',
  function_id: 'pipeline::produce',
  config: { api_path: '/pipeline/start', http_method: 'POST' },
})

File: skills/references/channels.py
"""
Pattern: Channels (Python)
Comparable to: Unix pipes, gRPC streaming, WebSocket data streams

Demonstrates binary streaming between workers: creating channels,
passing refs across functions, writing/reading binary data, and
using text messages for signaling.

How-to references:
  - Channels: https://iii.dev/docs/how-to/use-channels
"""

import json
import os

from iii import InitOptions, Logger, register_worker

engine_url = os.environ.get("III_ENGINE_URL", "ws://localhost:49134")
iii = register_worker(
    address=engine_url,
    options=InitOptions(worker_name="channels-example"),
)

# ---------------------------------------------------------------------------
# 1. Producer — creates a channel and streams data through it
# ---------------------------------------------------------------------------
async def produce(data):
    logger = Logger(service_name="pipeline::produce")

    # Create a channel pair
    channel = await iii.create_channel_async()

    # Pass the reader ref to the consumer via trigger
    await iii.trigger_async({
        "function_id": "pipeline::consume",
        "payload": {
            "reader_ref": channel.reader_ref,
            "record_count": len(data.get("records", [])),
        },
    })

    # Send metadata as a text message
    await channel.writer.send_message_async(
        json.dumps({"type": "metadata", "format": "ndjson", "encoding": "utf-8"})
    )

    # Stream records as binary data (newline-delimited JSON)
    for record in data.get("records", []):
        line = json.dumps(record) + "\n"
        await channel.writer.write(line.encode("utf-8"))

    # Signal end of stream
    await channel.writer.close_async()
    logger.info("Producer finished streaming", {"records": len(data.get("records", []))})

    return {"status": "streaming"}

iii.register_function("pipeline::produce", produce)

# ---------------------------------------------------------------------------
# 2. Consumer — receives a channel ref and reads the stream
# ---------------------------------------------------------------------------
async def consume(data):
    logger = Logger(service_name="pipeline::consume")

    # The reader ref is automatically resolved to a ChannelReader instance
    reader = data["reader_ref"]

    # Listen for text messages (metadata, signaling)
    messages = []
    reader.on_message(lambda msg: messages.append(json.loads(msg)))

    # Read entire binary stream
    raw = await reader.read_all()
    decoded = raw.decode("utf-8").strip()
    
    if not decoded:
        records = []
    else:
        lines = decoded.split("\n")
        records = [json.loads(line) for line in lines if line.strip()]

    logger.info("Consumer processed records", {"count": len(records)})
    return {"processed": len(records)}

iii.register_function("pipeline::consume", consume)

# ---------------------------------------------------------------------------
# 3. HTTP trigger to kick off the pipeline
# ---------------------------------------------------------------------------
iii.register_trigger({
    "type": "http",
    "function_id": "pipeline::produce",
    "config": {"api_path": "/pipeline/start", "http_method": "POST"},
})
